Numba

Ban-zee opened the pull request:
Numba has been kept optional. Added benchmarks tests for the modified methods(mainly variance, histogram and stats.circstd). These modified methods are twice as fast as the original numpy and scipy methods. Along with this, I have added a class in utils to allow the user to enable or disable numba at will. The required tests have been added as well.

canyon289 commented on arviz/stats/diagnostics.py:
```diff
@@ -510,13 +511,22 @@ def geweke(ary, first=0.1, last=0.5, intervals=20):
 
         z_score = first_slice.mean() - last_slice.mean()
-        z_score /= np.sqrt(first_slice.var() + last_slice.var())
+        if numba_check():
```
This comment is not required. While this if else pattern works I challenge us both to come up with a "nicer" pattern. I have some ideas and we can talk together about this.

Again don't worry about it right now but did want to make the comment

canyon289 replied:
@Ban-zee In the last 20 minutes I thought about this and think we should change on this PR before we move.

We shouldn't do a numba check everytime a function is called. I think a preferred pattern would be to check if Numba is installed once when ArviZ is loaded and then use that constant throughout the entire program.

I also think this will fit in well because users can optionally turn numba off, even if they have it installed.

If you want I can set some time to talk about this synchronously if you want :)

Ban-zee replied:
This can be done. I have a few questions in mind and I am up for discussion anytime. :)

canyon289 commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,84 @@
+
+
+class Hist:
+	def time_histogram(self):
```
These look like tabs not spaces :) 

Can you run black over this file for consistent formatting of all .py files?

canyon289 commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,84 @@
+
+
+class CircStd:
+	def time_circ_std(self):
```
Looks like its still tabs in here. I ask that this be converted to spaces

Ban-zee replied:
Sure, look's like something went wrong with black.

canyon289 commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,84 @@
+			return np.histogram(data, bins=100)
+
+
+class Variance:
```
Testing independent logic is a good start and makes it easy to see how numba should be implemented. Is it possible to also benchmark the arviz functions themselves with and without numba?

Ban-zee replied:
This is an issue. This was my initial plan but importing arviz breaks the build and the benchmarks do not execute. In my notebook, the reduction in the timings of the arviz methods was somewhat proportionate to the reduction in the time of the modified function; so I thought that if importing arviz does not work then I can benchmark the modified methods to ensure that the things are going in the right direction.

canyon289 replied:
Strategy makes sense. Let us know how we can help you figure out how to keep the build from breaking. Doesn't need to be done on this pr

canyon289 commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,86 @@
+class Hist:
+    def time_histogram(self):
+        try:
+            data = np.random.rand(10000, 1000)
```
Small thing. In these checks the data line is repeated but it seems it doesn't have to be

Ban-zee replied:
It can be changed. It's just that asv takes into account the data initialization time as well so I had to include it in both blocks so as to nullify this effect.

canyon289 replied:
Got it. Today I learned something

OriolAbril commented:
Could it be that `asv.conf.json` should be ignored and not added to ArviZ? It is full of local paths and `.asv/` was added to `.gitignore`

OriolAbril commented:
Also, out of curiosity, have you done any experiments or benchmarks with numba on `make_ufunc`?

Ban-zee commented:
I have removed the relative path; I had added it for my ease and forgot to replace it :| 
Regarding its inclusion, the conf file is needed for running the benchmarks locally. .asv contains environments and the results of the tests so I decided to gitignore it.
 
Unfortunately, I haven't tested `make_ufunc` yet :(.

OriolAbril commented on asv.conf.json:
```diff
@@ -0,0 +1,129 @@
+    //"install_timeout": 600,
+
+    // the base URL to show a commit for the project.
+    "show_commit_url": "https://github.com/Ban-zee/arviz/commit/",
```
Should this be modified to ArviZ instead of fork?

Ban-zee replied:
Don't know how I missed that :( , I'll change it by my next commit

canyon289 commented:
Just curious who is @ABM8? @Ban-zee is this another account that you have? 
<img width="313" alt="image" src="https://user-images.githubusercontent.com/7213793/59803040-364d4600-929f-11e9-8e7f-9956a25f4657.png">

Ban-zee commented:
That's me.

canyon289 commented:
> That's me. Looks like I did not rebase :(. I'll resolve this after I wake up.

Out of curiosity why not use one account?

Ban-zee commented:
It has stuck with me ever since I made the account and I did not bother changing it. :|

Ban-zee commented:
Just need to add one or two more tests for increasing the code coverage. Apart from that, the pr is up for review.

OriolAbril commented on arviz/stats/diagnostics.py:
```diff
@@ -13,10 +13,11 @@
-from ..utils import _var_names
-
+from ..utils import _var_names, conditional_jit, conditional_vect
+from .. import Numba
```
I _think_ that the cyclic import problem comes from this line. This line is in both diagnostics.py and stats.py, thus, it would explain why the cyclic import warnings finish always in one of these two files, and it looks like it is importing from _all_ ArviZ, hence the cycle. I don't really know how to import things from the __init__ file from within the same library though.

Ban-zee replied:
init too imports stats, plots and data. Hence the cyclic import I guess.

Ban-zee commented:
Up for review!!

OriolAbril commented on arviz/plots/elpdplot.py:
```diff
@@ -12,7 +12,8 @@
 )
-from ..stats import waic, loo, ELPDData
+from ..stats import waic, loo
+from ..stats import ELPDData
```
`from ..stats import waic, loo, ELPDData`

OriolAbril commented on arviz/tests/test_diagnostics.py:
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 from numpy.testing import assert_almost_equal, assert_array_almost_equal
 import pandas as pd
+import scipy.stats as st
```
`from scipy.stats import circstd`? I may have missed some call to st, but it looks like only circstd is used

OriolAbril commented on arviz/utils.py:
```diff
@@ -80,3 +80,58 @@ def wrapper(function):
+        if not cls.numba_flag and numba_check():
+            cls.numba_flag = True
+        else:
+            raise ValueError("Numba is already enabled or not installed")
```
Should an error be raised when calling `Numba.enable_numba()` if numba is already enables? It feels more natural to just leave numba activated and do nothing (like calling `%matplotlib inline` when inline backend is already set.

Ban-zee replied:
Wanted to play safe here but after looking at it again, I think this won't lead to any breakdown.

ahartikainen commented on arviz/stats/diagnostics.py:
```diff
@@ -445,6 +441,12 @@ def mcse(data, *, var_names=None, method="mean", prob=None):
 
+@conditional_vect
+def _sqr(a_a, b_b):
+    return math.sqrt(a_a + b_b)
```
Is math really needed? also name should be `_sqrt`?

    (a_a + b_b)**0.5

ahartikainen commented on arviz/stats/diagnostics.py:
```diff
@@ -859,22 +891,32 @@ def _conv_quantile(ary, prob):
-    sd = np.std(ary, ddof=1)
+    if _numba_flag:
+        ary = np.ravel(ary)
+        sd = np.sqrt(svar(ary, ddof=1))
```
_sqrt?

ary.ravel() to same line?

ahartikainen commented on arviz/stats/diagnostics.py:
```diff
@@ -859,22 +891,32 @@ def _conv_quantile(ary, prob):
-    sd = np.std(ary, ddof=1)
+    if _numba_flag:
+        ary = np.ravel(ary)
+        sd = np.sqrt(svar(ary, ddof=1))
```
Should there be ssd/sstd function?

Ban-zee replied:
I guess we can use _sqr, as you suggested.

ahartikainen commented on arviz/stats/stats_utils.py:
```diff
@@ -425,3 +426,30 @@ def __str__(self):
+        b_b = b_b + i * i
+    var = b_b / (len(data)) - ((a_a / (len(data))) ** 2)
+    var = var * (len(data) / (len(data) - ddof))
+    return var
```
Can we test this if there are huge and tiny values in the same array

ahartikainen commented on arviz/tests/test_diagnostics.py:
```diff
@@ -150,7 +152,7 @@ def test_deterministic(self):
         assert abs(np.median(reference["rhat_rank"] - arviz_data["rhat_rank"]) < 1e-14).all(None)
         not_rhat = [col for col in reference.columns if col != "rhat_rank"]
-        assert (abs(reference[not_rhat] - arviz_data[not_rhat]) < 1e-11).all(None)
+        assert (abs((reference[not_rhat] - arviz_data[not_rhat])).values < 1e-8).all(None)
```
Is this just random noise or did the function change?

Ban-zee replied:
There was a small change :( so I had to adjust the tests.

ahartikainen replied:
Ok. (The accuracy for rhat is something like 1e-6, so this change is fine)

ahartikainen commented on arviz/tests/test_diagnostics.py:
```diff
@@ -433,12 +459,12 @@ def test_multichain_summary_array(self, draws, chains):
+            assert_almost_equal(ess_mean_hat, ess_mean_hat_)
+            assert_almost_equal(ess_sd_hat, ess_sd_hat_)
+            assert_almost_equal(ess_bulk_hat, ess_bulk_hat_)
+            assert_almost_equal(ess_tail_hat, ess_tail_hat_)
```
Is there some difference between the results?

Ban-zee replied:
one of the results was the rounded version of the other one. This rounding off took place at around 10th or 11th place of the decimal iirc.

ahartikainen replied:
Can you check the summary code and see if it needs update? (It's is basically copy from the main functions, just trying to minimize duplicate calculations)

canyon289 commented on arviz/tests/test_diagnostics.py:
```diff
@@ -520,3 +557,75 @@ def test_split_chain_dims(self, chains, draws):
+
+
+def test_numba_bfmi():
+    state = Numba.numba_flag
```
All these test methods need docstrings

canyon289 commented on arviz/stats/diagnostics.py:
```diff
@@ -530,7 +543,11 @@ def ks_summary(pareto_tail_indices):
       Dataframe containing k diagnostic values.
     """
-    kcounts, _ = np.histogram(pareto_tail_indices, bins=[-np.Inf, 0.5, 0.7, 1, np.Inf])
+    _numba_flag = Numba.numba_flag
```
Can this be set at the top once instead of being called in every func?

ahartikainen replied:
Would that work?

Ban-zee replied:
It's only being called twice; will setting this it at the top make any difference?

canyon289 replied:
If its only being called twice then I don't think its worth it

canyon289 commented on arviz/stats/diagnostics.py:
```diff
@@ -510,13 +514,22 @@ def geweke(ary, first=0.1, last=0.5, intervals=20):
 
         z_score = first_slice.mean() - last_slice.mean()
-        z_score /= np.sqrt(first_slice.var() + last_slice.var())
+        if _numba_flag:
```
My next challenge for you is to get rid of all these if else statements.

There might be a pattern where we can do something like

```
z_score = _numba(f1, f2, args)
```

This way all the numba specific logic is contained in the Numba class and all the math logic is left here. What do you think?

Ban-zee replied:
Sure.

Ban-zee commented:
**Screenshot of the benchmark tests**
![bench-T](https://user-images.githubusercontent.com/29407246/59968430-79d1d980-9557-11e9-8f36-fc1551269485.png)

ColCarroll commented:
This is cool -- just so I'm clear about reading this, `numba` is making `circ_std` ~30% faster, `histogram` ~60% faster, and `variance` ~50% faster?

Ban-zee commented:
Approximately yes, on my machine.

Ban-zee commented:
It is a single element array

On Sun, 23 Jun 2019 2:05 am Ari Hartikainen <notifications@github.com wrote:

> *@ahartikainen* commented on this pull request.
>
> small comment on float formatting, not sure if its real issue or not
> ------------------------------
>
> In arviz/stats/diagnostics.py
> <https://github.com/arviz-devs/arviz/pull/710#discussion_r296450288>:
>
> > @@ -993,7 +990,7 @@ def _mc_error(ary, batches=5, circular=False):
>                      std = stats.circstd(ary, high=np.pi, low=-np.pi)
>              else:
>                  if _numba_flag:
> -                    std = np.sqrt(svar(ary))
> +                    std = float(_sqrt(svar(ary), np.zeros(1)))
>
> Float over array does not work
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arviz-devs/arviz/pull/710?email_source=notifications&email_token=AHALQDXF5QPVWODZT76TO2TP32EKTA5CNFSM4HY7OZ32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4LACDA#pullrequestreview-253100300>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHALQDQWRCBITGKQCPQOXI3P32EKTANCNFSM4HY7OZ3Q>
> .
>

ColCarroll commented:
`np.float_` for safety, or are you confident it will always be size 1?

ahartikainen commented:
But `_var` returns always float, right? then `np.zeros(1, dtype=float)`

Ban-zee commented:
_var does but _sqrt does not. It returns an nd array so the type conversion. The results will always be a single element as batched_traces is 2 dimensional

canyon289 commented on arviz/stats/diagnostics.py:
```diff
@@ -621,9 +638,9 @@ def _rhat(ary):
     chain_mean = np.mean(ary, axis=1)
     # Calculate chain variance
-    chain_var = np.var(ary, axis=1, ddof=1)
+    chain_var = _numba_var(svar, np.var, ary, axis=1, ddof=1)
```
This looks nice. Thank you!

canyon289 commented on arviz/utils.py:
```diff
@@ -100,3 +100,66 @@ def format_sig_figs(value, default=None):
+
+
+def _numba_var(function_1, function_2, data, axis=None, ddof=0):
+    """To replace the if-else blocks."""
```
Slight nudge here to change the function_1, function_2 to something more human recognizable like "numba_func", "numpy_func" or something like that, and expand out docstring

I would also change the docstring from 

`     """To replace the if-else blocks."""`
because this only makes sense to us from history. It won't make sense to someone in the future because they can't see the if else blocks that were replaced.

Something like 

> Returns appropriate function for numba speedup if Numba is installed or enabled

With argument documentation as well.

canyon289 commented on arviz/utils.py:
```diff
@@ -100,3 +100,66 @@ def format_sig_figs(value, default=None):
+        return wrapper
+
+
+def numba_check():
```
Can this be added to the Numba class to consolidate all the logic?

It doesn't have to be but just asking for stylistic and organization reasons

Ban-zee replied:
In my opinion, putting `numba_check` inside the class would not work as `numba_flag` is a class variable and calling class functions within a class is not possible in python. To get it working, I might have to switch to instance but that might lead to further breakdowns.

Ban-zee replied:
In my opinion, putting `numba_check` inside the class would not work as `numba_flag` is a class variable and calling class functions within a class is not possible in python. To get it working, I might have to switch to instance but that might lead to further breakdowns.

canyon289 replied:
Can class functions not be called within a class? The following code works, one class method calls the other. Are we thinking of the same thing?

```
class Test:
    @classmethod
    def say_hi(cls): 
        print("hi")

    @classmethod 
    def say_hi_for_real(cls):
        print("Really gonna say hi")
        cls.say_hi()
    
Test.say_hi_for_real()
```

> Really gonna say hi
> hi

Ban-zee replied:
My aim is to set `numba_flag` to `numba_check()` inside the class itself. 

    @classmethod
    def numba_check(cls):
        """Check if numba is installed."""
        numba = importlib.util.find_spec("numba")
        return numba is not None

    numba_flag = numba_check()

    @classmethod
    def disable_numba(cls):
        """To disable numba."""
        cls.numba_flag = False

    @classmethod
    def enable_numba(cls):
        """To enable numba."""
        if numba_check():
            cls.numba_flag = True
        else:
            raise ValueError("Numba is not installed")

Ban-zee replied:
I was aiming for something like this:
class Numba:


    numba_flag = None

    @classmethod
    def numba_check(cls):
        """Check if numba is installed."""
        numba = importlib.util.find_spec("numba")
        return numba is not None

    numba_flag = Numba.numba_check()

    @classmethod
    def disable_numba(cls):
        """To disable numba."""
        cls.numba_flag = False

    @classmethod
    def enable_numba(cls):
        """To enable numba."""
        if cls.numba_check():
            cls.numba_flag = True
        else:
            raise ValueError("Numba is not installed")


A possible workaround is the following :

    @classmethod
    def numba_check(cls):
        """Check if numba is installed."""
        numba = importlib.util.find_spec("numba")
        return numba is not None

    @classmethod
    def set_flag(cls):
        cls.numba_flag = cls.numba_check()

    @classmethod
    def disable_numba(cls):
        """To disable numba."""
        cls.numba_flag = False

    @classmethod
    def enable_numba(cls):
        """To enable numba."""
        if cls.numba_check():
            cls.numba_flag = True
        else:
            raise ValueError("Numba is not installed")

But this would lead extra code and tests which would defeat the purpose of including `numba_check` inside the class. 
I guess leaving `numba_check` outside the class should be fine.

canyon289 replied:
Ah I see. Alright no worries was just curious

Ban-zee commented:
What happened to coverage?

ahartikainen commented:
It's 96.72%

https://coveralls.io/builds/24205010

I have noticed that the following number is the only number to trust on that page.

![image](https://user-images.githubusercontent.com/13161958/60191730-797a5c80-983d-11e9-9eed-813e1c1ad42f.png)

canyon289 commented on arviz/tests/test_utils.py:
```diff
@@ -134,3 +151,83 @@ def placeholder_func():
+    assert flag == Numba.numba_flag
+
+
+@pytest.mark.parametrize("axis", (0, 1))
```
@Ban-zee As I get time Im going to keep picking with comments :)

Can we move numba disable to a parametrization? That way if the numba fails it's easier to tell because the tests are separated?

Ban-zee replied:
Sure

canyon289 commented:
It's my opinion that we should merge this sooner rather than later, before there's any merge conflicts.

@Ban-zee
Not required for this PR but I want to ask here

I may have missed this but is there documentation on how to run the speed tests?
Is it possible to add speed tests to ci?

Ban-zee commented:
How about if I add this to contributing: 
To run the **benchmark steps** do the following:

- Install airspeed velocity via `pip install asv`

- `cd arviz/`

- `asv run`

OriolAbril commented on CONTRIBUTING.md:
```diff
@@ -208,4 +208,14 @@ Alternatively, to start a jupyter notebook, there are two steps, first run:
+
+- `cd arviz/`
+
+- `asv run`
```
what about?

    $ cd arviz/
    $ asv run

minor comment, it is only to follow the convention from the rest of the contributing file.

Ban-zee replied:
Sure, and I'll change ELPD data too

OriolAbril commented on arviz/stats/diagnostics.py:
```diff
@@ -453,6 +448,12 @@ def mcse(data, *, var_names=None, method="mean", prob=None):
+
+
+@conditional_jit
 def geweke(ary, first=0.1, last=0.5, intervals=20):
```
I have no idea why nor how to solve it, I point it out just in case. If you build the docs locally, you will see that the `geweke` link on the api page is not working anymore. It shows the text, but it does not link to the `geweke` page.

canyon289 replied:
Oh I think we might need to do a functools wrap. Decorators "eat" the docstrings of their decorated functions

https://www.thecodeship.com/patterns/guide-to-python-function-decorators/

canyon289 replied:
I think we should do this on this PR because we don't want to regress the docs for every function that we use `conditional_jit` for

OriolAbril commented on doc/index.rst:
```diff
@@ -66,6 +66,7 @@ Here is the citation in BibTeX format
     Quickstart<notebooks/Introduction>
     Example Gallery<examples/index>
     Cookbook<notebooks/InferenceDataCookbook>
+    Numba<notebooks/Numba>
```
Completely unrelated. Should this toctree have the same elements as the sticky nav-bar on top of the website? 

If so this seems like a good ocasion to add `InferenceData<notebooks/XarrayforArviZ>` here and `("Numba", "notebooks/Numba"),` between lines 137-138 of `doc/conf.py`. Adding numba to the nav-bar is probably more important, otherwise it can be hard to find.

Ban-zee commented:
Strange, the tests work fine.

ahartikainen commented:
````
with pytest.raises(ValueError) as err:
    get_coords(data, coords)
    
>       assert "Coords {'NOT_A_COORD_NAME'} are invalid coordinate keys" in str(err)

arviz/tests/test_plot_utils.py:112: AssertionError
````

Is says AssertationError is not ValueError. Did pytest update?

Ban-zee commented:
Sorry for the delay :)
